packages/components/src/components/Combobox/Combobox.tsx

```diff
@@ -0,0 +1,148 @@
+    customStore?: AriakitComboboxProps["store"];
+  };
+
+const matchSorter = (
+  list: ComboboxItemProps[],
+  value: string,
+): ComboboxItemProps[] => {
+  return filter(list, (item) => includes(toLower(item.label), toLower(value)));
+};
+
+const getFilteredItems = <T extends ComboboxItemProps>(
+  items: T[],
+  searchValue: string,
+  filterItem?: boolean | ((inputValue: string, item: T) => boolean),
+): T[] => {
+  if (isBoolean(filterItem)) {
+    return filterItem
+      ? (matchSorter(items as ComboboxItemProps[], searchValue) as T[])
+      : items;
+  }
+  if (isFunction(filterItem)) {
+    return filter(items, (item) => filterItem(searchValue, item));
+  }
+  return items;
+};
+
+// eslint-disable-next-line jsdoc/require-param
+/**
+ * Combobox combines a text input and a popover to allow users to filter from long lists of options.
+ */
+const ComboboxInner = <T extends ComboboxItemProps>(
+  {
+    items,
+    onChangeInput,
+    onSelectItem,
+    placeholder = "Select an option",
+    noResultsMessage = "No results found",
+    filterItem = true,
+    hasError,
+    customStore,
+    ...props
+  }: ComboboxProps<T>,
+  ref: React.ForwardedRef<HTMLInputElement>,
+): React.JSX.Element => {
+  const store = useComboboxStore({
+    store: customStore,
+  });
+  const selectedValue = store.getState().selectedValue;
+  const [searchValue, setSearchValue] = useState("");
+  const [selectedLabel, setSelectedLabel] = useState<string | undefined>("");
+
+  const filteredItems = getFilteredItems(items, searchValue, filterItem);
+
+  return (
+    <ComboboxProvider
+      setSelectedValue={(value) => onSelectItem?.(value)}
+      setValue={(value) => {
+        const label = find(items, ["value", value])?.label;
+        setSearchValue(value);
+        setSelectedLabel(label);
+      }}
+      store={store}
+      value={selectedLabel}
+    >
+      <ComboboxInput
+        aria-label="Select an option"
+        hasError={hasError}
+        onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
+          onChangeInput?.(event);
+          setSelectedLabel(event.target.value);
+        }}
+        placeholder={placeholder}
+        ref={ref}
+        store={store}
+        textOverflow="ellipsis"
+        {...props}
+      />
+      <ComboboxPopover sameWidth>
+        {filteredItems.length === 0 ? (
+          <ComboboxItem>{noResultsMessage}</ComboboxItem>
+        ) : (
+          map(filteredItems, (item) => (
+            <ComboboxItem
+              aria-selected={selectedValue === item.value}
+              data-value={item.value}
+              disabled={item.disabled}
+              key={item.value}
+              value={item.value}
+            >
+              {item.label}
+            </ComboboxItem>
+          ))
+        )}
+      </ComboboxPopover>
+    </ComboboxProvider>
+  );
+};
+
+const Combobox = forwardRef(ComboboxInner) as <T extends ComboboxItemProps>(
```

Comment:
Initially, when I opened the PR, I forgot to pass the type `T` so we could use it in the apps. When I tried to fix it here, I had some issues when using forwardRef. I managed to fix it by using the following as an example: https://gist.github.com/IrvingArmenta/a6d7fc76ed538697ad18b7f074accdde.